File: api/dense/solve/compute_qtb.m
function bs = compute_qtb(Q, bs, ncols) 
%#codegen -args {coder.typeof(0, [inf, inf], [1,1]), coder.typeof(0, [inf,inf], [1,1]), int32(0)}
nrow = int32(size(Q,1));
for k=1:ncols
    % Optimized version for
    % bs(k:nrow,:) = bs(k:nrow,:) - 2*v*(v'*bs(k:nrow,:)),
    % where v is Q(k:npngs)
    for jj=1:int32(size(bs,2))
        t2 = 0; 
        for ii=k:nrow; t2 = t2+Q(ii,k)*bs(ii,jj); end; 
        t2 = t2+t2;
        for ii=k:nrow; bs(ii,jj) = bs(ii,jj) - t2 * Q(ii,k); end
    end
end

File: api/dense/solve/codegen/lib/compute_qtb/compute_qtb.c
#include "compute_qtb.h"

void compute_qtb(const emxArray_real_T *Q, emxArray_real_T *bs, int ncols)
{
  int nrow;
  int k;
  int i;
  int jj;
  double t2;
  int ii;
  nrow = Q->size[0];
  for (k = 0; k < ncols; k++) {
    i = bs->size[1];
    for (jj = 0; jj < i; jj++) {
      t2 = 0.0;
      for (ii = k + 1; ii <= nrow; ii++) {
        t2 += Q->data[(ii + Q->size[0] * k) - 1] * bs->data[(ii + bs->size[0] *
          jj) - 1];
      }

      t2 += t2;
      for (ii = k + 1; ii <= nrow; ii++) {
        bs->data[(ii + bs->size[0] * jj) - 1] -= t2 * Q->data[(ii + Q->size[0] *
          k) - 1];
      }
    }
  }
}

void compute_qtb_initialize(void)
{
}

void compute_qtb_terminate(void)
{
}
